Source/JavaScriptCore/builtins/SetPrototype.js

```diff
@@ -44,3 +44,254 @@ function forEach(callback /*, thisArg */)
         callback.@call(thisArg, key, key, this);
     } while (true);
 }
+
+function union(other)
+{
+    "use strict";
+
+    if (!@isSet(this))
+        @throwTypeError("Set operation called on non-Set object");
+
+    // Get Set Record
+    if (!@isObject(other))
+        @throwTypeError("Set.prototype.union expects first parameter to be an object");
+    var size = @toNumber(other.size);
+    // size is NaN
+    if (size !== size)
+        @throwTypeError("Set.prototype.union expects other.size to be a non-NaN number");
+
+    var has = other.has;
+    if (!@isCallable(has))
+        @throwTypeError("Set.prototype.union expects other.has to be callable");
+
+    var keys = other.keys;
+    if (!@isCallable(keys))
+        @throwTypeError("Set.prototype.union expects other.keys to be callable");
+
+    var iterator = keys.@call(other, keys);
+    var wrapper = {
+        @@iterator: function () { return iterator; }
+    };
+
+    var result = @setClone(this);
+    for (var key of wrapper)
+        result.@add(key);
+
+    return result;
+}
+
+function intersection(other)
+{
+    "use strict";
+
+    if (!@isSet(this))
+        @throwTypeError("Set operation called on non-Set object");
+
+    // Get Set Record
+    if (!@isObject(other))
+        @throwTypeError("Set.prototype.intersection expects first parameter to be an object");
+    var size = @toNumber(other.size);
+    // size is NaN
+    if (size !== size)
+        @throwTypeError("Set.prototype.intersection expects other.size to be a non-NaN number");
+
+    var has = other.has;
+    if (!@isCallable(has))
+        @throwTypeError("Set.prototype.intersection expects other.has to be callable");
+
+    var keys = other.keys;
+    if (!@isCallable(keys))
+        @throwTypeError("Set.prototype.intersection expects other.keys to be callable");
+
+    var result = new @Set();
+    if (this.@size <= size) {
+        var bucket = @setBucketHead(this);
+
+        do {
+            bucket = @setBucketNext(bucket);
+            if (bucket === @sentinelSetBucket)
+                break;
+            var key = @setBucketKey(bucket);
+            if (has.@call(other, key))
+                result.@add(key);
+        } while (true);
+    } else {
+        var iterator = keys.@call(other, keys);
+        var wrapper = {
+            @@iterator: function () { return iterator; }
+        };
+
+        for (var key of wrapper) {
+            if (this.@has(key))
+                result.@add(key);
+        }
```

Comment:
With the current spec, you'll actually need to add to a temporary set and then iterate over the receiver in order to ensure that the elements remain in their original order. This was going to be my comment:
```
// The idea of the current spec is that:
// 1. We want to have the output elements stay in their original order.
// 2. We don't want to perform a greater number of user-observable has() calls on the argument set
//    than the number of elements in it.
//
// So, we iterate over the argument set when it's the smaller one, but the underlying assumption is
// that we *won't* need to iterate over the receiver set in order to sort the keys. In SM & V8,
// "deterministic hash tables" allow the ordering of two elements to be checked by a pointer comparison,
// but our implementation (with a linked list of buckets) does not allow for this.
// Accordingly, the spec may change here.
```
I was discussing this with Kevin yesterday, as I [shared on Slack](https://webkit.slack.com/archives/CTV4FGWF4/p1675219659716579?thread_ts=1675150621.902089&cid=CTV4FGWF4), and it sounds like he'll reraise this to committee in the near future.

Reply:
Oof, I guess I missed that step when I read the spec. I agree that's not great for our implementation... For now I'll just put a fixme in the code and in the option.